### DISCUSSION TITLE
How to yank releases on private index?

### DISCUSSION THREAD

**Discussion by lochsh** [devpi/devpi#1037] [Q&A]
We run an instance of `devpi-server` internally to act as a pypi.org mirror and to publish our internal packages to for internal users at our company.

We had an issue with some recent internal releases, and would like to yank the associated files. I couldn't find any specific documentation from devpi about how to do this. Trying to use `devpi patchjson` hasn't been fruitful -- I consistently get 404s (I am able to use `getjson` on the same resource successfully). For example, if I do `patchjson` with a JSON file containing the contents of the `result` attribute of the JSON returned from `getjson` I get a 404.

If using `devpi patchjson` is a suitable way to do this, what would you expect to be in the JSON file argument? Any examples in tests have an un-nested dictionary in the JSON for updating a single top-level attribute. If the result of `devpi getjson /user/index/package` is:

```javascript
{
    "0.0.0": {
        "+links": [
            {
                 "hash_spec": "...",
                 "href": "http://server:3141/user/index/.../package-0.0.0.whl",
            },
            ...
        ],
        ...
    },
    "0.1.0": ...
}
```

and I want to yank 0.0.0, I assume I'd want:

```diff
             {
                  "hash_spec": "...",
                  "href": "http://server:3141/user/index/.../package-0.0.0.whl",
+                 "data-yanked": "Broken wheel",
             },
             ...
         ],
```

is that correct? Apologies if I'm missing something obvious.

Should I be using a JSON patch? I had trouble with that too. Here is an example:

```javascript
[
 { "op": "add", "path": "/+links/0/data-yanked", "value": "Wheel missing shared library" },
 { "op": "add", "path": "/+links/1/data-yanked", "value": "Wheel missing shared library" },
 { "op": "add", "path": "/+links/2/data-yanked", "value": "Wheel missing shared library" },
 { "op": "add", "path": "/+links/3/data-yanked", "value": "Wheel missing shared library" },
 { "op": "add", "path": "/+links/4/data-yanked", "value": "Wheel missing shared library" },
 { "op": "add", "path": "/+links/5/data-yanked", "value": "Wheel missing shared library" }
]
```

used with `devpi patchjson /user/index/package/0.0.0/ updates.json` also resulted in a 404 response.

Thanks in advance for any help!

**Comment by fschulze** [comment]
Your line of thinking was correct. Unfortunately it doesn't work. The keys with a ``+`` in front are generated from other data and can't be changed with ``patchjson``. This is not documented properly yet.

Currently there is no way to yank releases. The next major release will allow that, but will also require an export/import cycle, as the data storage will change to allow for features like this.
